### ISSUE TITLE
implement `expect { }.toThrow` as assertion function

### ISSUE THREAD

**Issue by robstoll** [robstoll/atrium#97]
*Platform* (JVM, JS and/or Android): all

## Code related feature
Expecting that an assertion is thrown is currently implemented aside from other assertion functions. The reason for it is that it is using a different assertion verb and a dynamic representation of the subject (depending on whether an exception was thrown or not). 
We could already define it for `Expect<out () -> Any?>` but then the error message would look something like the following which does not tell us what we would want to see.
```
expect: () -> kotlin.Int
- throws: IllegalStateException
```
vs:
```
expect the thrown Exception: no exception was thrown
- is an instance of type: IllegalStateException
```

We could provide a way to change the assertion verb on `ReportingAssertionContainer`. That's the type which is actually returned by `expect` but only for the very first call (afterwards it is `Expect`).
Well... then we still use something special to implement it. 

I guess if we want to implement it as normal assertion function, then we cannot circumvent that the first line is `expect: () -> kotlin.Int` but we could then do something like:
```
expect: () -> kotlin.Int
-> thrown exception when called: no exception thrown
   >> is instance of type: IllegalStateException
```

Following a PoC based on `Expect` and the new SubjectChanger and FeatureExtractor:
```
inline fun <reified T : Any> Expect<out () -> Any?>.toThrow(): Expect<T> {
    return ExpectImpl.changeSubject.unreported(this) {
        try {
            it()
            null
        } catch (e: Throwable) {
            e
        }
    }.let { expectThrowable ->
        ExpectImpl.feature.extractor(expectThrowable)
            .withDescription(Untranslatable("thrown exception when called"))
            .withRepresentationForFailure(Untranslatable("no exception occurred"))
            .withCheck { it != null }
            .withFeatureExtraction { it as Throwable }
            .withoutSubAssertions()
            .build()
            .isA<T>()
    }
}
```
This is not ideal yet as `isA` should also be reported in case no exception is thrown but we can see that it would be feasible.

The benefit of having such a function:
- we only have to define two assertion verb functions: one without assertionCreator lambda and the other which delegates to the first and uses `addAssertionsCreatedBy`
- the assertion function could be used in arbitrary places -> e.g. say one wants to be sure that an exception is thrown after a certain call
- we can get rid of ThrowableThrown.Builder and reuse existing functionality

**Comment by robstoll** [comment]
I currently don't see a use case where I would use such a check within a sub-assertion. Therefore I labelled it with `need votes`. Please vote for it if you need it, I won't implement it otherwise

**Comment by robstoll** [comment]
implemented with 7e09509
